pysmt/solvers/msat.py

```diff
@@ -48,6 +47,88 @@
 from pysmt.walkers.identitydag import IdentityDagWalker
 
 
+class _MSatEnvWrapper(object):
+    "A simple wrapper for msat_env objects with explicit reference counting"
+
+    __slots__ = ['msat_env', 'refs']
+    
+    def __init__(self, msat_env):
+        self.msat_env = msat_env
+        self.refs = 1
+
+    def ref(self):
+        self.refs += 1
+        return self
+
+    def unref(self):
+        assert self.refs > 0
+        self.refs -= 1
+        if not self.refs:
+            mathsat.msat_destroy_env(self.msat_env)
+
+    def __call__(self):
+        return self.msat_env
+
```

Comment:
I think this can be simplified by adding a ```__del__``` method in _MSatEnvWrapper, and let python deal with the ref-counting. If every object that uses the msat_env keeps an internal copy (as it seems to be the case), everything should work.